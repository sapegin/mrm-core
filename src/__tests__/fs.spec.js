jest.mock('fs');
jest.mock('../util/log', () => ({
	added: jest.fn(),
	removed: jest.fn(),
}));

const fs = require('fs-extra');
const vol = require('memfs').vol;
const log = require('../util/log');
const _fs = require('../fs');
const readFile = _fs.readFile;
const updateFile = _fs.updateFile;
const copyFiles = _fs.copyFiles;
const deleteFiles = _fs.deleteFiles;
const makeDirs = _fs.makeDirs;

const fs$copySync = fs.copySync;
const fs$removeSync = fs.removeSync;
const fs$ensureDirSync = fs.ensureDirSync;

afterEach(() => {
	vol.reset();
	log.added.mockClear();
	log.removed.mockClear();
});

describe('readFile()', () => {
	it('should read a file', () => {
		const contents = 'test';
		vol.fromJSON({ '/a': contents });

		const result = readFile('/a');

		expect(result).toBe(contents);
	});

	it('should strip BOM marker', () => {
		const contents = 'test';
		vol.fromJSON({ '/a': '\uFEFF' + contents });

		const result = readFile('/a');

		expect(result).toBe(contents);
	});
});

describe('updateFile()', () => {
	it('should update a file', () => {
		const contents = 'test';
		vol.fromJSON({ '/a': contents });

		updateFile('/a', 'pizza', contents, true);

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should create a file', () => {
		updateFile('/a', 'pizza', 'test');

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should create a folder', () => {
		updateFile('/a/b', 'pizza', 'test');

		expect(vol.toJSON()).toMatchSnapshot();
	});
});

describe('copyFiles()', () => {
	afterEach(() => {
		fs.copySync = fs$copySync;
	});

	it('should copy a file', () => {
		vol.fromJSON({ '/tmpl/a': 'pizza' });

		copyFiles('/tmpl', 'a');

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should copy multiple files', () => {
		vol.fromJSON({ '/tmpl/a': 'pizza', '/tmpl/b': 'coffee' });

		copyFiles('/tmpl', ['a', 'b']);

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should not try to copy a file if contents is the same', () => {
		fs.copySync = jest.fn();

		vol.fromJSON({ '/tmpl/a': 'pizza', '/tmpl/b': 'pizza', '/a': 'pizza', '/b': 'coffee' });

		copyFiles('/tmpl', ['a', 'b']);

		expect(fs.copySync).toHaveBeenCalledTimes(1);
		expect(fs.copySync).toBeCalledWith('/tmpl/b', 'b', {});
	});

	it('should not overwrite a file if overwrite=false', () => {
		const json = { '/tmpl/a': 'pizza', '/a': 'coffee' };
		vol.fromJSON(json);

		copyFiles('tmpl', 'a', { overwrite: false });

		expect(vol.toJSON()).toEqual(json);
	});

	it('should throw when source file not found', () => {
		const fn = () => copyFiles('tmpl', 'a');

		expect(fn).toThrowError('source file not found');
	});

	it('should print a file name', () => {
		vol.fromJSON({ '/tmpl/a': 'pizza' });

		copyFiles('/tmpl', 'a');

		expect(log.added).toBeCalledWith('Copy a');
	});

	it('should not print a file name if contents is the same', () => {
		vol.fromJSON({ '/tmpl/a': 'pizza', '/a': 'pizza' });

		copyFiles('/tmpl', 'a');

		expect(log.added).toHaveBeenCalledTimes(0);
	});
});

describe('deleteFiles()', () => {
	afterEach(() => {
		fs.removeSync = fs$removeSync;
	});

	it('should delete a file', () => {
		vol.fromJSON({ '/a': 'pizza', '/b': 'coffee' });

		deleteFiles('/a');

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should delete multiple files', () => {
		vol.fromJSON({ '/a': 'pizza', '/b': 'coffee', '/c': 'schawarma' });

		deleteFiles(['/a', '/b']);

		expect(vol.toJSON()).toMatchSnapshot();
	});

	it('should not try to delete a file if it doesnâ€™t exist', () => {
		fs.removeSync = jest.fn();

		vol.fromJSON({ '/a': 'pizza' });

		deleteFiles(['/a', '/b']);

		expect(fs.removeSync).toHaveBeenCalledTimes(1);
		expect(fs.removeSync).toBeCalledWith('/a');
	});

	it('should not throw when file not found', () => {
		const fn = () => deleteFiles('/a');

		expect(fn).not.toThrowError();
	});

	it('should print a file name', () => {
		vol.fromJSON({ '/a': 'pizza' });

		deleteFiles('/a');

		expect(log.removed).toBeCalledWith('Delete /a');
	});

	it('should not print a file name if file not found', () => {
		deleteFiles('/a');

		expect(log.removed).toHaveBeenCalledTimes(0);
	});
});

describe('makeDirs()', () => {
	afterEach(() => {
		fs.ensureDirSync = fs$ensureDirSync;
	});

	it('should create a folder', () => {
		makeDirs('/a');

		expect(fs.statSync('/a').isDirectory()).toBe(true);
	});

	it('should create multiple folders', () => {
		makeDirs(['/a', '/b']);

		expect(fs.statSync('/a').isDirectory()).toBe(true);
		expect(fs.statSync('/b').isDirectory()).toBe(true);
	});

	it('should not try to create a folder if it exists', () => {
		fs.ensureDirSync = jest.fn();

		vol.fromJSON({ '/a/1': 'pizza' });

		makeDirs(['/a', '/b']);

		expect(fs.ensureDirSync).toHaveBeenCalledTimes(1);
		expect(fs.ensureDirSync).toBeCalledWith('/b');
	});

	it('should print a folder name', () => {
		makeDirs('/a');

		expect(log.added).toBeCalledWith('Create folder /a');
	});

	it('should not print a folder name if folder exists', () => {
		vol.fromJSON({ '/a/b': 'pizza' });

		deleteFiles('/a');

		expect(log.added).toHaveBeenCalledTimes(0);
	});
});
