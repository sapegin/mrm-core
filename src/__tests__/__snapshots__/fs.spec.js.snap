// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copyFiles() should copy a file 1`] = `
Object {
  "/a": "pizza",
  "/tmpl/a": "pizza",
}
`;

exports[`copyFiles() should copy multiple files 1`] = `
Object {
  "/a": "pizza",
  "/b": "coffee",
  "/tmpl/a": "pizza",
  "/tmpl/b": "coffee",
}
`;

exports[`copyFiles() should not try to copy a file if contents is the same 1`] = `
Object {
  "/a": "pizza",
  "/b": "pizza",
  "/tmpl/a": "pizza",
  "/tmpl/b": "pizza",
}
`;

exports[`deleteFiles() should delete a file 1`] = `
Object {
  "/b": "coffee",
}
`;

exports[`deleteFiles() should delete multiple files 1`] = `
Object {
  "/c": "schawarma",
}
`;

exports[`updateFile() should create a file 1`] = `
Object {
  "/a": "pizza",
}
`;

exports[`updateFile() should create a folder 1`] = `
Object {
  "/a/b": "pizza",
}
`;

exports[`updateFile() should update a file 1`] = `
Object {
  "/a": "pizza",
}
`;
